storage/innobase/buf/buf0buf.cc

```diff
@@ -429,6 +429,29 @@ buf_pool_get_oldest_modification(void)
 	return(oldest_lsn);
 }
 
+ulint
+buf_get_flush_list_len(buf_pool_t *buf_pool) {
+	ulint i;
+	ulint pages = 0;
+	if (buf_pool == NULL) {
+		for (i = 0; i < srv_buf_pool_instances; i++) {
+			buf_pool_t *buf_pool_instance;
+
+			buf_pool_instance = buf_pool_from_array(i);
+
+			buf_flush_list_mutex_enter(buf_pool_instance);
```

Comment:
Not a strong opinion here, but do we need to acquire this mutex here?
```#define UT_LIST_GET_LEN(BASE) (BASE).count```
So even if we dirty-read non-zero value, we will return zero next time, which will cause one additional loop turn (line 3485).

No other comments than from @satya-bodapati

Reply:
@kamil-holubicki @altmannmarcelo what if n_pending_reads is 0 and the flush_list_len len is also seen 0 (after a dirty page is added)? Isn't this possible?

Reply:
@satya-bodapati I don't think this is possible, at this stage of shutdown, the only thing that is running is purge and the pending io reads. as explained on the above workflow, we increment n_pend_reads -> add page to flush_list so it's now possible to have a new page added if n_pend_reads == 0

Reply:
@altmannmarcelo I mean, due to instruction (load/store) reordering between n_pending_reads & flush_list.count. These are not atomics and no read/write barriers here.

Reply:
just realized I pinged the wrong Satya - cc @satya-bodapati

Reply:
@altmannmarcelo right, once pending_iops is 0, a page is already added to the flush list just before the last sweep. 
The flush coordinator thread at the same moment can see the total number of dirty pages to be 0. this is because the other thread (incremented it) but the current thread still sees old value due to the missing read barrier?

So instead of seeing count as 1, page cleaner coordinator thread can see the cached value of flush list count as 0? :thinking: isn't it?

Reply:
> The flush coordinator thread at the same moment can see the total number of dirty pages to be 0. this is because the other thread (incremented it) but the current thread still sees old value due to the missing read barrier?

@satya-bodapati here the same thread (io thread) will be adding the page to flush list, merge all ibuf records to this age, then once it is completed, it will mark the page as IO_BUF_NONE (at this point PC will consider the page eligible for flushing) and only then decrement the number of pending reads.

IMHO for the purposes of the last PC sweep, we do not need to have the read of flush_list under mutex as pending reads will only be zero once the page is already eligible for flushing.

To be honest, here we are not under too much multithreading here - IO threads are only processing the async IO that have been previously dispatched to them, no new operations will land (only the ones resulting from completing this IO, such ibuf), user threads have already been closed at this point. So even adding the mutex back will not mean too much overheard.